pkg/cmd/variable/set/http.go

```diff
@@ -0,0 +1,156 @@
+package set
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/cli/cli/v2/api"
+	"github.com/cli/cli/v2/internal/ghrepo"
+	"github.com/cli/cli/v2/pkg/cmd/variable/shared"
+)
+
+const (
+	createdOperation = "Created"
+	updatedOperation = "Updated"
+)
+
+type setPayload struct {
+	Name         string  `json:"name,omitempty"`
+	Repositories []int64 `json:"selected_repository_ids,omitempty"`
+	Value        string  `json:"value,omitempty"`
+	Visibility   string  `json:"visibility,omitempty"`
+}
+
+type setOptions struct {
+	Entity        shared.VariableEntity
+	Environment   string
+	Key           string
+	Organization  string
+	Repository    ghrepo.Interface
+	RepositoryIDs []int64
+	Value         string
+	Visibility    string
+}
+
+type setResult struct {
+	Err       error
+	Key       string
+	Operation string
+}
+
+func setVariable(client *api.Client, host string, opts setOptions) setResult {
+	var err error
+	var postErr api.HTTPError
+	result := setResult{Operation: createdOperation, Key: opts.Key}
+	switch opts.Entity {
+	case shared.Organization:
+		if err = postOrgVariable(client, host, opts.Organization, opts.Visibility, opts.Key, opts.Value, opts.RepositoryIDs); err == nil {
+			return result
+		} else if errors.As(err, &postErr) && postErr.StatusCode == 409 {
+			// Server will return a 409 if variable already exists
+			result.Operation = updatedOperation
+			err = patchOrgVariable(client, host, opts.Organization, opts.Visibility, opts.Key, opts.Value, opts.RepositoryIDs)
+		}
+	case shared.Environment:
+		var ids []int64
+		ids, err = api.MapReposToIDs(client, opts.Repository.RepoHost(), []ghrepo.Interface{opts.Repository})
+		if err != nil || len(ids) != 1 {
+			err = fmt.Errorf("failed to look up repository %s: %w", ghrepo.FullName(opts.Repository), err)
+			break
+		}
```

Comment:
Not sure if this should just be passed in `setOptions` rather than doing the look up here.

Reply:
Do not think there is much difference either way. Accessing id here before calling the API might be a little better as we may make less calls in event of input validation failure/any other error.